site/docs/plugins/conversations.md

```diff
@@ -7,1190 +7,1501 @@ next: false
-You also have to install the conversations plugin itself, before you can register individual conversations on your bot.
+If you perform any API calls via `ctx.api` (including `ctx.reply`), the plugin takes care of them automatically.
+In contrast, your own database communcation needs special treatment.
 
-```ts
-// Install the session plugin.
-bot.use(session({
-  initial() {
-    // return empty object for now
-    return {};
-  },
-}));
+This is done as follows.
 
-// Install the conversations plugin.
-bot.use(conversations());
-```
+### The Golden Rule of Conversations
 
-Next, you can install the conversation builder function as middleware on your bot object by wrapping it inside `createConversation`.
+Now that [we know how conversations are executed](#conversations-are-replay-engines), we can define one rule that applies to the code you write inside a conversation builder function.
+You must follow it if you want your code to behave correctly.
 
-```ts
-bot.use(createConversation(greeting));
-```
+::: warning THE GOLDEN RULE
+
+**Code behaving differently between replays must be wrapped in `conversation.external`.**
 
-Now that your conversation is registered on the bot, you can enter the conversation from any handler.
-Make sure to use `await` for all methods on `ctx.conversation`---otherwise your code will break.
+:::
+
+This is how to apply it:
 
 ```ts
-bot.command("start", async (ctx) => {
-  await ctx.conversation.enter("greeting");
-});
+// BAD
+const response = await accessDatabase();
+// GOOD
+const response = await conversation.external(() => accessDatabase());
 ```
 
-As soon as the user sends `/start` to the bot, the conversation will be entered.
-The current context object is passed as the second argument to the conversation builder function.
-For example, if you start your conversation with `await ctx.reply(ctx.message.text)`, it will contain the update that contains `/start`.
+Escaping a part of your code via `conversation.external` signals to the plugin that this part of the code should be skipped during replays.
+The return value of the wrapped code is stored by the plugin and reused during subsequent replays.
+In the above example, this prevents repeated database access.
 
-::: tip Change the Conversation Identifier
-By default, you have to pass the name of the function to `ctx.conversation.enter()`.
-However, if you prefer to use a different identifier, you can specify it like so:
+USE `conversation.external` when you ...
 
-```ts
-bot.use(createConversation(greeting, "new-name"));
-```
+- read or write to files, databases/sessions, the network, or global state
+- call `Math.random()` or `Date.now()`
+- perform API calls on `bot.api` or other independent instances of `Api`
+
+DO NOT USE `conversation.external` when you ...
 
-In turn, you can enter the conversation with it:
+- call `ctx.reply` or `ctx.api.sendMessage` or similar methods
+- use any other JavaScript syntax like functions, classes, if-else, loops, etc
```

Comment:
I'm a little confused about this part. In the previous example, you can use `Math.random()`, `Date.now()`, or `accessDatabase()`, but here we shouldn't use functions, classes, or any other JavaScript syntax. Isn't that contradictory?

Reply:
Oh. I see the problem. I'll rephrase this. Thanks.

What I meant was that if you define your own function or use any JS syntax, you don't need to wrap this. But maybe this is more confusing than helping.